### DISCUSSION TITLE
How to get crossed edges for a node in cost::TransitionCost?

### DISCUSSION THREAD

**Discussion by grzezlo** [valhalla/valhalla#4636] [Q&A] [answered]
Is there any way of getting `tile` from `node` or `edge` inside the `TransitionCost` method without modifying this method's parameters list in all costing models and all calls of it?
I'd like to experiment with improving BlindUserMode to compose route through crosswalks instead of intersections directly.
When we have some intersection with a crosswalk drawn near-by, Valhalla is often generating route directly through the intersection.
One solution would be to decrease `walkway_factor`, but it leads to non-optimal results in other situations.
Other, better, would be to add some penalty for the situation when road is intersecting an other road, leading to preferring situations when a road is intersected by a footway (a crosswalk).
In cases, where there is no crosswalk drawn near given roads intersection, this intersection itself should be chosen by the algorithm due to lack of other route variants.
So my concept is to modify `PedestrianCost::TransitionCost` like this:

```
if (BlindUserMode && !shortest_ && edge->is_road()
    && IsRoadsIntersection(node, edge) {
  // Increase the cost by penalty
}
      ```

`IsRoadsIntersection` would return true if at least two edges other than a analysed edge, are roads too.
As I understand, to get list of edges from a given node, I must have a tile of this node.
And there is no other way of getting the tile, than passing it from the `Thor`, which would require extending parameters list in all implementations of `TransitionCost` and `TransitionCostReverse`, am I right?

**Comment by kevinkreiser** [comment, marked as answer]
yes! for the sake of testing you can do the following:

1. add a new method to DynamicCost.h that allows you to attach a GraphReader to it
2. at the top of whichever algorithm you are using (bidirection astar probably) call this new method to give your costing object access to the graphreader
3. from inside TransitionCost you have the predecessor edgelabel. this object had the "endnode" id which you can use to get the graphtile that the node is in from the graphreader you now have access to

**Reply by grzezlo** [reply]
Great idea, thank you @kevinkreiser, I'll try this way. 
But why should I use the `endnode` of `pred`, not `edge` or simply the `node` parameter itself?
For adding costs for intersection in `PedestrianCost::TransitionCost` the `edge` object is used, not `pred`.
```
if (edge->edge_to_right(idx) && edge->edge_to_left(idx)) 
```

**Reply by kevinkreiser** [reply]
well the node parameter is a pointer and you dont know what tile it belongs to. to get the tile that has the node and the edges connected to it you can use the pred.endnode. it was just an example though. feel free to ask more question about how the data is structured, its not 100% intuitive on first glance :smile:

**Comment by grzezlo** [comment]
Thanks @kevinkreiser, it worked and Valhalla routes tested by us are better  near crossings now!
Next things not related with intersections, are further BlindUserMode improvements we'd like to test, e.g. 

- Adding path_factor to discourage edges with use::kPath, which often are hard to travel by blind users and sometimes non-existent in terrain (when a meadow became a building after years).
- Adding maneuver_penalty for pedestrian (type_== kBlind && !shortest_) to encourage routes with less turns and other maneuvers, so generated routes should be simpler to travel.

No questions here, implementation will be simple, so I'm writing about it just for your information.
